Freetype: to CMake

johnwparent opened the pull request:
Convert Freetype to use CMake to support Windows

part of #34938

johnwparent committed:
freetype - multi build system

free type style

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -3,10 +3,12 @@
 
 
-class Freetype(AutotoolsPackage):
+class Freetype(CMakePackage):
```
Shouldn't this include `AutotoolsPackage`? Even when adding it, `autotools` does not show up as a valid choice for `build_system`, though:
```
==> Error: invalid values for variant "build_system" in package "freetype": ['autotools']
```

johnwparent replied:
CMake is the buildsystem for this package now. We no longer want to inherit autotools at that point.

michaelkuhn replied:
Does it hurt to keep Autotools? It's still Freetype's default build system.

johnwparent replied:
I don't personally see utility in maintaining two build systems when one should suffice and provides improved cross platform support. While Spack does support a package having multiple builders, I think avoiding it where possible is ideal.

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -47,6 +48,13 @@ def headers(self):
+        return [self.define("FT_WITH_ZLIB", True), self.define("FT_WITH_BZIP2", True)]
+
+
+class AutoToolsBuilder(AutotoolsBuilder):
```
Not sure if it makes a difference: `Autotools` instead of `AutoTools`?

johnwparent replied:
This has been removed all together in the latest revisions.

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -47,6 +48,13 @@ def headers(self):
+
+class CMakeBuilder(CMakeBuilder):
+    def cmake_args(self):
+        return [self.define("FT_WITH_ZLIB", True), self.define("FT_WITH_BZIP2", True)]
```
1. Why is `FT_WITH_ZLIB` true? It's explicitly disabled for Autotools. We also do not have a `zlib` dependency.
2. What about the other options? This should probably enable/disable the same features as the Autotools variant.

johnwparent replied:
This was a mistake I made during a (very old) previous iteration. It was _supposed_ to be `FT_DISABLE_ZLIB` set to ON. This is fixed on the current state.

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -30,7 +32,6 @@ class Freetype(AutotoolsPackage):
 
     depends_on("bzip2")
     depends_on("libpng")
-    depends_on("pkgconfig", type="build")
```
Why is this not required anymore?

johnwparent replied:
This will not be required as we are not longer using autotools.

michaelkuhn replied:
CMake still seems to use pkg-config, though: https://github.com/freetype/freetype/blob/master/CMakeLists.txt#L251

johnwparent replied:
It's certainly not a hard requirement as I am able to build perfectly fine without it on both Unix and Windows, and as evidenced by the lack of `REQUIRED` in the find package invocation. Many find package modules may use PkgConfig to perform some package detection, but it's not required, and the only direct use I see is the `pkg_check_modules` which comes from the CMake module so it'll be available anyway. I'll make the requirement optional and off on Windows and on everywhere else, I think that's probably best, as you're right, other platforms may get some utility out of it (although I'm not sure how much)

michaelkuhn replied:
Perfect, thanks!

johnwparent commented:
@michaelkuhn This was taken from another PR. Looks like I didn't cherry-pick enough history, let me grab the rest of it, I think you'll find the port is much better.

johnwparent committed:
Rest of freetype history - squashed

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -47,14 +47,13 @@ def headers(self):
+            self.define("FT_DISABLE_ZLIB", True),
+            self.define("FT_DISABLE_BROTLI", True),
+            self.define("FT_DISABLE_HARFBUZZ", True),
+            self.define("FT_DISABLE_PNG", True),
```
Why do we disable PNG? We used to enable it and still have a `libpng` dependency.

johnwparent replied:
That was me reading the wrong line. Good catch.

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -47,14 +47,13 @@ def headers(self):
+            self.define("FT_DISABLE_BROTLI", True),
+            self.define("FT_DISABLE_HARFBUZZ", True),
+            self.define("FT_DISABLE_PNG", True),
+            self.define("FT_WITH_BZIP2", True),
```
Are you sure this is correct?
```
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_TESTING
    FT_WITH_BZIP2
```

johnwparent replied:
~~I took it from freetype's cmake file, let me double check.~~ My mistake, you're right this is not the correct variable.

michaelkuhn commented on var/spack/repos/builtin/packages/freetype/package.py:
```diff
@@ -47,14 +47,13 @@ def headers(self):
+            self.define("FT_WITH_BZIP2", True),
         ]
-        if self.spec.satisfies("@2.9.1:"):
-            args.append("--enable-freetype-config")
```
It seems the CMake version is missing `freetype-config`. We should keep the same behavior.

johnwparent replied:
Perhaps you can elaborate here as I'm not particularly familiar with Freetype's build system but would that not be the flag for generating the config files that CMake generates by default?  Otherwise I cannot find a reference to the config files in the CMake interface. If this is indeed not a true paralell, you can disregard my comment regarding dropping autotools altogether, as you're right we don't want to drop functionality.

michaelkuhn replied:
This flag causes a `freetype-config` binary to be generated that can be used similarly to `pkg-config freetype`. I think some of Freetype's dependents might need it.

johnwparent replied:
After digging into things a bit, CMake properly exports this project and produces a freetype config that in general should be sufficient for dependent projects, but I found a few that truly required the autotools style config that can be treated like a binary so I'm adding back the autotools build. When/If those packages are ever ported to be solely cmake projects, this can be just a cmake package, but until then, autotools will need to be an option on unix.

johnwparent replied:
So feel free to disregard my "removing autotools entirely" comment and related thoughts, as they are wrong.

johnwparent committed:
restore autotools

freetype-config has an alternative in the cmake build system
but some packages require autotools produced version
restore autotools to preserve option for projects that need it.

johnwparent committed:
Add build system directive

johnwparent commented:
@spackbot run pipeline

michaelkuhn reviewed (approved):
Thanks for your work on this!

johnwparent commented:
@spackbot run pipeline

michaelkuhn commented:
@spackbot run pipeline
